File: Managed/NextTurn.UE.Runtime/Slate/HorizontalBoxSlot.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using NextTurn.UE.Annotations;

namespace Unreal.Slate
{
    public class HorizontalBoxSlot : Slot
    {
        private unsafe NativeHorizontalBoxSlot* Target => (NativeHorizontalBoxSlot*)this.pointer;

        public unsafe int HorizontalAlignment
        {
            get => this.Target->HorizontalAlignment;
            set => NativeMethods.SetHorizontalAlignment(this.Target, value);
        }

        public unsafe int VerticalAlignment
        {
            get => this.Target->VerticalAlignment;
            set => NativeMethods.SetVerticalAlignment(this.Target, value);
        }

        private struct NativeHorizontalBoxSlot
        {
            public int HorizontalAlignment { get; }

            public int VerticalAlignment { get; }
        }

        private static class NativeMethods
        {
            [Calli]
            public static extern unsafe void SetHorizontalAlignment(NativeHorizontalBoxSlot* slot, int horizontalAlignment);

            [Calli]
            public static extern unsafe void SetVerticalAlignment(NativeHorizontalBoxSlot* slot, int verticalAlignment);
        }
    }
}

File: Source/NextTurnRuntime/Private/SlateCore/HorizontalBoxSlot.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#if WITH_APPLICATION_CORE

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/SlateCore/Public/Widgets/SBoxPanel.h"

ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME HorizontalBoxSlot

TYPE_EXPORT_START

void EXPORT_CALL_CONV SetHorizontalAlignment(SHorizontalBox::FSlot& Slot, EHorizontalAlignment HorizontalAlignment)
{
	Slot.HAlign(HorizontalAlignment);
}

void EXPORT_CALL_CONV SetVerticalAlignment(SHorizontalBox::FSlot& Slot, EVerticalAlignment VerticalAlignment)
{
	Slot.VAlign(VerticalAlignment);
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(SetHorizontalAlignment),
	EXPORT_METHOD(SetVerticalAlignment),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

#endif
